### DISCUSSION TITLE
How to use PCI?

### DISCUSSION THREAD

**Discussion by ChocolateLoverRaj** [phil-opp/blog_os#1360] [Questions]
In the [OS I'm making](https://github.com/ChocolateLoverRaj/code-runner/tree/edition-3), I want to add PCI (be able to at least list the PCI devices). I parsed the ACPI table with the acpi library, but did not successfully use APIC for interrupts instead of pic8259. What is the best (hopefully minimal) example of using PCI in a Rust OS?

**Comment by tpdenk** [comment]
https://github.com/tsatke/devos/blob/cf9d2ff1ca1ca973372e6dd15a3ad8a589adf73e/kernel/src/driver/pci/raw.rs

I'm not sure whether you really need ACPI for PCI.

Make sure you don't confuse `ACPI`, `APIC` and `PCI`. Those are three completely different things.

**Reply by ChocolateLoverRaj** [reply]
From https://wiki.osdev.org/PCI#Enumerating_PCI_Buses it says
> For both BIOS and UEFI systems, you can check the ACPI tables to determine if the memory mapped access mechanism is supported. 

The example you sent seems to use "Configuration Space Access Mechanism #1". Is there an example which uses the ACPI info to make sure that memory-mapped access is supported and then enumerates devices with the memory-mapped method?

Edit: I realized the osdev article says:
> Note that systems that do provide the memory mapped access mechanism are also required to support PCI access mechanism #1 for backwards compatibility. 

and
> For UEFI systems, it's extremely safe to assume that mechanism #2 is not supported

so the example should work for modern computers

**Reply by ChocolateLoverRaj** [reply]
Also are there any libraries to use PCI in your own OS similar to how there are `acpi` and `ps2-mouse` libraries? Are there any in-progress libraries?

**Reply by bjorn3** [reply]
The pci_types crate is useful IMO.
